### PR TITLE
[WIP][SS] RocksDB state store changelog checkpointing

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
### What changes were proposed in this pull request?
Implement changelog checkpointing for rocksdb state store provider.

### Why are the changes needed?
make checkpoint latency predictable

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1940,6 +1940,19 @@ object SQLConf {
       // 5 is the default table format version for RocksDB 6.20.3.
       .createWithDefault(5)
 
+  /**
+   * Note: this is defined in `RocksDBConf.FORMAT_VERSION`. These two places should be updated
+   * together.
+   */
+  val STATE_STORE_ROCKSDB_CHANGE_CHECKPOINTING_ENABLED =
+    buildConf("spark.sql.streaming.stateStore.rocksdb.enableChangelogCheckpointing")
```
Comment (anishshri-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1944]: Maybe rephrase as `rocksdb.changelogCheckpoint.enabled` ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+import org.apache.commons.io.IOUtils
+import org.apache.hadoop.fs.{FSError, Path}
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.io.CompressionCodec
+import org.apache.spark.sql.execution.streaming.CheckpointFileManager
+import org.apache.spark.sql.execution.streaming.CheckpointFileManager.CancellableFSDataOutputStream
+
+
+class ChangelogReader(fm: CheckpointFileManager, fileToRead: Path,
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: Can we rename as `RocksDBStateStoreChangelogReader` ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import com.google.common.io.ByteStreams
+import org.apache.commons.io.IOUtils
+import org.apache.hadoop.fs.{FSError, Path}
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.io.CompressionCodec
+import org.apache.spark.sql.execution.streaming.CheckpointFileManager
+import org.apache.spark.sql.execution.streaming.CheckpointFileManager.CancellableFSDataOutputStream
+
+
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: Nit: newline and can we add a class comment ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+          val valueBuffer = new Array[Byte](valueSize)
+          ByteStreams.readFully(input, valueBuffer, 0, valueSize)
+          byteArrayPair = new ByteArrayPair(keyBuffer, valueBuffer)
+        }
+      }
+    }
+  }
+}
+
+class ChangelogWriter(fm: CheckpointFileManager, file: Path,
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: Same as above ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala [hunk 5]
```diff
@@ -195,6 +195,12 @@ private[sql] class RocksDBStateStoreProvider
     new RocksDBStateStore(version)
   }
 
+  override def getReadStore(version: Long): StateStore = {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:198]: Hmm, how was this working before this change for RocksDB ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 6]
```diff
@@ -718,7 +781,8 @@ object RocksDBConf {
       getBooleanConf(BOUNDED_MEMORY_USAGE_CONF),
       getLongConf(MAX_MEMORY_USAGE_MB_CONF),
       getRatioConf(WRITE_BUFFER_CACHE_RATIO_CONF),
-      getRatioConf(HIGH_PRIORITY_POOL_RATIO_CONF))
+      getRatioConf(HIGH_PRIORITY_POOL_RATIO_CONF),
+      storeConf.compressionCodec)
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:788]: Where do we check for valid/supported codecs here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 7]
```diff
@@ -150,8 +161,24 @@ class RocksDB(
...
+                put(byteArrayPair.key, byteArrayPair.value)
+              } else {
+                remove(byteArrayPair.key)
+              }
+            }
+          } finally {
+            if (changelogReader != null) changelogReader.close()
+          }
+        }
+        numKeysOnLoadedVersion = numKeysOnWritingVersion
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:184]: Is `numKeys` being updated somewhere here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 8]
```diff
@@ -150,8 +161,24 @@ class RocksDB(
           metadata.numKeys
         }
         numKeysOnWritingVersion = numKeys
-        numKeysOnLoadedVersion = numKeys
 
+        for (v <- latestSnapshotVersion + 1 to version) {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:166]: Do we need this logic if changelog checkpointing is disabled too ?

Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:166]: Maybe add a small comment for what this snippet is doing ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    compressedStream.write(key)
+    compressedStream.writeInt(value.size)
+    compressedStream.write(value)
+  }
+
+  def delete(key: Array[Byte]): Unit = {
+    assert(compressedStream != null)
+    compressedStream.writeInt(key.size)
+    compressedStream.write(key)
+    compressedStream.writeInt(-1)
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: Curious, why are we adding this ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+      // Closing the compressedStream causes the stream to write/flush flush data into the
+      // rawStream. Since the rawStream is already closed, there may be errors.
+      // Usually its an IOException. However, Hadoop's RawLocalFileSystem wraps
+      // IOException into FSError.
+      case e: FSError if e.getCause.isInstanceOf[IOException] =>
+
+      // SPARK-42668 - Catch and log any other exception thrown while trying to cancel
+      // raw stream or close compressed stream.
+      case NonFatal(ex) =>
+        logInfo(s"Failed to cancel delta file for provider " +
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: Can we also log the delta file name/path ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala (hunk 2 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: This can actually be applied to hdfs state store, but we may not have the time to do the refactoring there.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala (hunk 9 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: A -1 key length signals the end of files, which is the same format as hdfs state store changelog.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala (hunk 10 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: Sure.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala (hunk 2 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: Ok - if its going to apply to both, then you can call it `StateStoreChangeLogReader` ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala (hunk 5 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:198]: It loads a write state store, but the load doesn't consume extra resources. We make the change here because now we create the changelog writer in load.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala (hunk 3 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: Makes sense.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala (hunk 9 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: Right - but for delete, its trying to signal like empty value ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 11]
```diff
@@ -286,44 +318,44 @@ class RocksDB(
...
-    val checkpointDir = createTempDir("checkpoint")
-    var rocksDBBackgroundThreadPaused = false
+    assert(newVersion > 0)
     try {
-      // Make sure the directory does not exist. Native RocksDB fails if the directory to
-      // checkpoint exists.
-      Utils.deleteRecursively(checkpointDir)
 
       logInfo(s"Flushing updates for $newVersion")
       val flushTimeMs = timeTakenMs { db.flush(flushOptions) }
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:327]: Are we not making the periodic flush changes in this PR ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala (hunk 2 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: Sounds good.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala (hunk 9 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala]: No, for delete, we have | key length | key bytes | value length = -1 |, 
key length is non-negative for both put and delete, a -1 in the key length field is end of file.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 8 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:166]: It will be a noop without changelog checkpointing because latest snapshot will be loaded successfully.

Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:166]: Sure.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 7 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:184]: Yes, during changelog replay, numKeysOnWritingVersion is updated inside put() and remove().

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 11 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:327]: I personally feel this pr is large enough, so prefer to add it as a followup improvements.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (chaoqin-li1123) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1944]: Ok with that, don't have strong opinion about naming here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 6 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:788]: There is no check for codec in HDFS backed state store also. It will throw exception when the first chanelog writer is initialized.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 8 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:166]: Probably still better to just embed in an `if` condition for better readability ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 6 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:788]: hmm, maybe we can fix it in both places then ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala [hunk 12]
```diff
@@ -134,6 +137,27 @@ class RocksDBFileManager(
...
   }
+  private val onlyDeltaFiles = new PathFilter {
+    override def accept(path: Path): Boolean = path.toString.endsWith(".changelog")
+  }
+
+  private lazy val sparkConf = Option(SparkEnv.get).map(_.conf).getOrElse(new SparkConf)
+
+  private def codec = CompressionCodec.createCodec(sparkConf, codecName)
+
+  def getChangeLogWriter(version: Long): StateStoreChangelogWriter = {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:148]: Why not just move these functions to `RocksDB.scala` itself and pass the reader/writer as required ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+        } else {
+          val valueBuffer = new Array[Byte](valueSize)
+          ByteStreams.readFully(input, valueBuffer, 0, valueSize)
+          // A put record.
+          byteArrayPair = new ByteArrayPair(keyBuffer, valueBuffer)
+        }
+      }
+    }
+  }
+}
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala]: Missing newline at end of file ?
